### ISSUE TITLE
Question: Why is negation limited to atoms

### ISSUE THREAD

**Issue by Gozala** [RhizomeDB/PomoLogic#3]
I have been wondering why is negation limited to atoms

https://github.com/RhizomeDB/PomoLogic/blob/e0a2b383cd5f08d7b950cf79147f95fe2bd47c90/README.md?plain=1#L221-L229

More specifically is there reason why negation can to be applied to the Rule's ? I realize you could write a negated rule, but it seems like it would be very useful means of abstraction for the query language.

 Also relatedly is there reason why rule can not have another rule application in it's own body ? It's certainly possible to inline (non-recursive) rules, yet it seems very useful means of abstraction.

**Comment by expede** [comment]
@Gozala I think you may be experiencing a namespace collision with Datalog terminology (these are not Lisp atoms, which is confusing for folks with that background). An atom may represent a fact or the head of a rule — by analogy to traditional languages, it's the function name and arguments but not its body.

A rule by contrast is the entire AST of the rule, including the body.

https://github.com/RhizomeDB/PomoLogic/blob/e0a2b383cd5f08d7b950cf79147f95fe2bd47c90/README.md?plain=1#L34-L46

**Comment by Gozala** [comment]
😅 you are right, I was initially confusing it with lisp atoms, but then after reading a section about it

https://github.com/RhizomeDB/PomoLogic/blob/e0a2b383cd5f08d7b950cf79147f95fe2bd47c90/README.md?plain=1#L34-L46

I got a wrong impression that it was unit that described multiple attribute, value relations on the same entity

Specifically in datomic terms I thought it was roughly equivalent of

```clj
(def foo
   '[:find [?v1 ?v2 ?vn]
     :where [?e :a1 v1]
            [?e :a2 v2]
            [?e :an vn]])
```

Thanks for responding, and I'm glad behavior my expectation here.

Am I correct to assume that `point(x: X, y: Y)` is roughly equivalent of this in datomic, unless there is a `point` rule defined already ?

```clj
[:find [?x ?y]
 :where [e? :x ?x]
        [e? :y ?y]]
```

**Comment by expede** [comment]
> Am I correct to assume that point(x: X, y: Y) is roughly equivalent of this in datomic, unless there is a point rule defined already ?

## Rhzome in Particular

TL;DR yes exactly

## Broader Theory

In the broader context (and a lot of the text about Datalog), it's more complicated. It depends if you mean in Datalog broadly or in Rhizome in particular. It may be helpful to expand on this in case you run into similar text elsewhere.

The example you give above is like an anonymous function. A _generic_ Datalog has different tables for different shapes of facts (relations). Not all Datalogs are restricted to the EAV shape. They can have `Point(x, y)` which would have to be queried in the form `?- Point(X, Y).`, which is different from `?- Pair(X, Y).`, despite them having the same arguments.

In Datomic, this can be emulated as something like this:

```
[:find [x? ?y]
 :where [e? :x ?x]
        [e? :y ?y]
        [e? :Relation :Point]]
```

**Comment by expede** [comment]
Also in case it's helpful for research, the notation that Datomic uses is [EDN](https://github.com/edn-format/edn) (pronounced like in "Garden of _Eden_")

**Comment by Gozala** [comment]
```edn
[:find [x? ?y]
 :where [e? :x ?x]
        [e? :y ?y]
        [e? :Relation :Point]]
```

Ah this is interesting, so it appears that the name of the atom is relevant. I assumed names were only for the human reader.

Are the use of names (in the context of PomoDB) a deliberate choice arrived after some consideration or just an inheritance carried over from prior systems ?

The reason I'm asking is because I have been implementing a query engine where I made following choices and I wonder if they were the right ones.

1. Names are for humans only and `point(x: 1, y: 2)` is no different from `pair(x: 1, y: 2)`.
    > Perhaps this wont come as a surprise to you given how many times I brought not wanting to have type names in IPLD schema and wanting something akin of [unison](https://www.unison-lang.org/docs/the-big-idea/)

   - One undoubtedly will want a `newtype` or what unison calls (in my opinion more clearly) a [unique type](https://www.unison-lang.org/docs/fundamentals/data-types/unique-and-structural-types/) but that is so easy here you define you type entity and then associate entity (which in datomic syntax would look like)

     ```edn
     [{:db/id Point :relation :Point}
      {:x 1 :y 2 :type Point}]
      ```

2. Entity identifiers are derived from the seed (attributes). This way it does not matter where the entity was created it will have a same identifier. If you want to create `point(x: 1, y: 2)` and structurally same yet semantically different entity, you can use causal reference to the first one. If you want to tie entity to the origin that is also straightforward (although can't decide if it should be a causal reference or an attribute)

I'd be love a feedback your feedback for the above and curious if something similar had been considered in the context of PomoDB.

**Comment by Gozala** [comment]
BTW above two idea also combine into making `:db/id` an obsolete construct as you could simply create transactions like these

```edn
[{:x 1 :y 2 :type {:relation :Point}
 {:x 2 :y 3 :type {:relation :Point}]
```

And because the same seed `{:relation :Point}` both points points would assert exact same entity for `{:relation :Point}` and correspondingly reference to that entity from both `:type` attribute.

**Comment by expede** [comment]
Oh yeah, probably worth noting that Rhizome (explicitly) uses Merklization and causality instead of centralised DB transaction autonumbering

**Comment by expede** [comment]
> Ah this is interesting, so it appears that the name of the atom is relevant. I assumed names were only for the human reader.

Yep!

> Are the use of names (in the context of PomoDB) a deliberate choice arrived after some consideration or just an inheritance carried over from prior systems ?

Named relations aren't a thing for "ground facts" in PomoDB, but you could emulate relations on top of that. Having named queries is probably a good idea. Take everything in these repos with a grain of salt — we never went through a deep design review and parts of it may be inconsistent from various iterations
